File: java_refactoring/src/main/java/org/example2/solution/TeaService.java
package org.example2.solution;

public abstract class TeaService {
    private final int weight;
    private double amountMade = 0;
    private double teaStrength = 0;
    private double teaFlavor = 0;

    protected TeaService(int weight) {
        if (weight <= 0) {
            throw new IllegalArgumentException("Tea weight must be positive.");
        }
        this.weight = weight;
    }

    public final void makeTea() {
        makeTeaType();
        if (teaStrength / 5 > 50) {
            System.out.println("Tea on sale made");
            System.out.println("Saved $1 on tea on sale");
        }
    }

    public final TeaPourResult pourTea(Cup cup) {
        System.out.println(cup.describe());
        pourTeaType();
        return new TeaPourResult(getTeaType(), amountMade);
    }

    public final TeaTaxInfo createTaxInfo() {
        return new TeaTaxInfo(getTeaType(), amountMade);
    }

    protected abstract TeaType getTeaType();

    protected abstract void makeTeaType();

    protected abstract void pourTeaType();

    protected final int getWeight() {
        return weight;
    }

    protected final double getTeaStrength() {
        return teaStrength;
    }

    protected final double getTeaFlavor() {
        return teaFlavor;
    }

    protected final void addAmountMade(double amount) {
        amountMade += amount;
    }

    protected final void addTeaStrength(double amount) {
        teaStrength += amount;
    }

    protected final void addTeaFlavor(double amount) {
        teaFlavor += amount;
    }
}

File: typescript_refactoring/example2/solution/TeaService.ts
import { Cup, TeaPourResult, TeaTaxInfo, TeaType } from "./TeaTypes";

export abstract class TeaService {
    protected amountMade = 0;
    protected teaStrength = 0;
    protected teaFlavor = 0;

    protected constructor(protected readonly weight: number) {
        if (weight <= 0) {
            throw new Error("Tea weight must be positive.");
        }
    }

    public makeTea(): void {
        this.makeTeaType();
        if (this.teaStrength / 5 > 50) {
            console.log("Tea on sale made");
            console.log("Saved $1 on tea on sale");
        }
    }

    public pourTea(cup: Cup): TeaPourResult {
        this.validateCup(cup);
        console.log(`${cup.type} ${cup.color} ${cup.material} ${cup.sizeMl}ml ${cup.weightGrams}g`);
        this.pourTeaType();

        return {
            teaType: this.teaType,
            amountRemaining: this.amountMade,
        };
    }

    public createTaxInfo(): TeaTaxInfo {
        return {
            teaType: this.teaType,
            salesTax: this.amountMade * 2.25,
            propertyTax: this.amountMade * 2.35,
            incomeTax: this.amountMade * 2.13,
            totalTax: this.amountMade * 2.68,
        };
    }

    protected abstract readonly teaType: TeaType;

    protected abstract makeTeaType(): void;

    protected abstract pourTeaType(): void;

    private validateCup(cup: Cup): void {
        if (cup.sizeMl <= 0 || cup.weightGrams <= 0) {
            throw new Error("Cup size and weight must be positive.");
        }
    }
}
